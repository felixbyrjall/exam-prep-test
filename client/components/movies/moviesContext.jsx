import React from "react";

const MoviesContext = React.createContext({
    fetchMovies: async () => [],
    onSaveMovie: async (movie) => {
    },
});