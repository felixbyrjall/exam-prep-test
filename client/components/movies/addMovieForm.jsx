import React, { useContext, useState } from "react";
import { MoviesContext } from "./moviesContext";

export function AddMovieForm() {
  const [title, setTitle] = useState("");
  const { postNewMovie } = useContext(MoviesContext);

  async function handleSubmit(e) {
    e.preventDefault();

    const movie = { title };
    await postNewMovie(movie);
  }

  return (
    <form onSubmit={handleSubmit}>
      <h2>Add movie</h2>
      <div>
        <input
          name={"title"}
          value={title}
          onChange={(e) => setTitle(e.target.value)}
        />
      </div>
      <div>
        <button>Submit</button>
      </div>
    </form>
  );
}
