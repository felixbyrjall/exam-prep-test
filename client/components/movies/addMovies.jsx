import React, { useContext, useState } from "react";
import { MoviesContext } from "./moviesContext";
import { useNavigate } from "react-router-dom";

export function AddMovies() {
  const [title, setTitle] = useState("");
  const [year, setYear] = useState("");
  const [metacritic, setMetacritic] = useState("");
  const navigate = useNavigate();
  const { postNewMovie } = useContext(MoviesContext);

  async function handleSubmit(e) {
    e.preventDefault();
    postNewMovie({ title, year, metacritic }).then(() => {
      navigate("/movies");
    });
  }

  return (
    <form onSubmit={handleSubmit}>
      <h2>Add movie</h2>
      <div>
        <div>
          TITLE:
          <input
            required
            name={"title"}
            value={title}
            onChange={(e) => setTitle(e.target.value)}
          />
        </div>
        <div>
          YEAR:
          <input
            required
            name={"year"}
            value={year}
            onChange={(e) => setYear(e.target.value)}
          />
        </div>
        <div>
          METACRITIC:
          <input
            required
            name={"metacritic"}
            value={metacritic}
            onChange={(e) => setMetacritic(e.target.value)}
          />
        </div>
      </div>
      <div>
        <button>Submit</button>
      </div>
    </form>
  );
}
