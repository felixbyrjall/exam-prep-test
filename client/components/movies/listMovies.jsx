import React, { useEffect, useState } from "react";
import { MoviesContext } from "./moviesContext";

export function ListMovies() {
  const [movies, setMovies] = useState([]);
  const [loading, setLoading] = useState(true);
  const [error, setError] = useState();

  function postNewMovie(movie) {
    setMovies((oldMovies) => [...oldMovies, movie]);
  }
  async function fetchMovies() {
    const res = await fetch("/api/movies");
    return await res.json();
  }

  async function loadMovies() {
    setLoading(true);
    setMovies([]);
    setError(undefined);

    try {
      setMovies(await fetchMovies());
    } catch (err) {
      setError(err);
    } finally {
      setLoading(false);
    }
  }

  useEffect(() => {
    loadMovies();
  }, []);

  return (
    <>
      <MoviesContext.Provider
        value={{
          fetchMovies,
          postNewMovie,
        }}
      >
        <h2>List the movies</h2>
        {loading && <div>Spinner</div>}
        {movies.map((m) => (
          <div key={m._id}>
            <h4>
              {m.title} ({m.year}) - Metacritic: {m.metacritic}
            </h4>
          </div>
        ))}
      </MoviesContext.Provider>
    </>
  );
}
