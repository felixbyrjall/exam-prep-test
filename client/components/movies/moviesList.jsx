import React, {useContext, useEffect, useState} from "react";
import {MoviesContext} from "./moviesContext";

export function MoviesList() {
    const [movies, setMovies] = useState([])
    const {fetchMovies} = useContext(MoviesContext)
    async function loadMovies(){
        setMovies(await fetchMovies());
    }

    useEffect(() => {
        loadMovies();
    }, []);

    return (
        <>
            <h2>List movies</h2>
            {movies.map((m) => (
                <div>{m.title}</div>
            ))}
        </>
    )
}