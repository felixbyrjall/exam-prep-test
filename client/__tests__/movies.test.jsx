import renderer, { act } from "react-test-renderer";
import { MoviesList } from "../components/movies/moviesList";
import { MoviesContext } from "../components/movies/moviesContext";
import { AddMovieForm } from "../components/movies/addMovieForm";
describe("movies react application", () => {
  it("shows loading screen", () => {
    const component = renderer.create(<MoviesList />);
    expect(component).toMatchSnapshot();
  });

  it("shows loaded movies", async () => {
    // snapshot test
    const fetchMovies = () => [
      { title: "Jest, the test", _id: 1 },
      { title: "Bob, the builder", _id: 2 },
    ];
    let component;
    await act(async () => {
      component = renderer.create(
        <MoviesContext.Provider value={{ fetchMovies }}>
          <MoviesList />
        </MoviesContext.Provider>,
      );
    });

    expect(component).toMatchSnapshot();
  });

  it("add a new movie", async () => {
    // A test that adds a new movie
    const postNewMovie = jest.fn();
    const component = renderer.create(
      <MoviesContext.Provider value={{ postNewMovie }}>
        <AddMovieForm />
      </MoviesContext.Provider>,
    );
    await act(async () => {
      component.root.findByType("input").props.onChange({
        target: { value: "My test movie" },
      });
    });
    await act(async () => {
      component.root.findByType("form").props.onSubmit({
        preventDefault() {},
      });
    });

    expect(postNewMovie).toBeCalledWith({
      title: "My test movie",
    });
  });
});
