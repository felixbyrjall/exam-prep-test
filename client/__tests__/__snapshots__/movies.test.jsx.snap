// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`movies react application shows loaded movies 1`] = `
[
  <h2>
    List the movies
  </h2>,
  <div>
    Jest, the test
  </div>,
  <div>
    Bob, the builder
  </div>,
]
`;
